src/Compilers/Core/Portable/SourceGeneration/Nodes/InputNode.cs

```diff
@@ -5,61 +5,61 @@
     /// </summary>
-    /// <typeparam name="T"></typeparam>
+    /// <typeparam name="T">The type of the input</typeparam>
     internal sealed class InputNode<T> : IIncrementalGeneratorNode<T>
     {
-        public NodeStateTable<T> UpdateStateTable(DriverStateTable.Builder graphState, NodeStateTable<T> previousTable, CancellationToken cancellationToken)
-        {
-            // the input node doesn't change the table. 
-            // instead the driver manipulates the previous table to contain the current state of the node.
-            // we can just return that state, as it's already up to date.
-            return previousTable;
-        }
-
-        // PROTOTYPE(source-generators): how does this work? we definitely want to be able to add custom comparers to the input nodes
-        // I guess its just a 'compare only' node with this as the input?
-        public IIncrementalGeneratorNode<T> WithComparer(IEqualityComparer<T> comparer) => this;
+        private readonly Func<DriverStateTable.Builder, ImmutableArray<T>> _getInput;
+        private readonly IEqualityComparer<T>? _comparer;
 
-        public NodeStateTable<T> CreateInputTable(NodeStateTable<T> previousTable, T item)
+        public InputNode(Func<DriverStateTable.Builder, ImmutableArray<T>> getInput, IEqualityComparer<T>? comparer = null)
         {
-            // PROTOTYPE(source-generators): we should compare the values, not just assume they were added
-            return NodeStateTable<T>.WithSingleItem(item, EntryState.Added);
+            _getInput = getInput;
+            _comparer = comparer;
         }
 
-        public NodeStateTable<T> CreateInputTable(NodeStateTable<T> previousTable, IEnumerable<T> items)
+        public NodeStateTable<T> UpdateStateTable(DriverStateTable.Builder graphState, NodeStateTable<T> previousTable, CancellationToken cancellationToken)
         {
+            var inputItems = _getInput(graphState);
+
             // create a mutable hashset of the new items we can check against
-            PooledHashSet<T> itemsSet = PooledHashSet<T>.GetInstance();
-            foreach (var item in items)
+            HashSet<T> itemsSet = new HashSet<T>(_comparer);
+            foreach (var item in inputItems)
             {
                 itemsSet.Add(item);
             }
 
-            var builder = new NodeStateTable<T>.Builder();
+            var builder = previousTable.ToBuilder();
 
             // for each item in the previous table, check if its still in the new items
             foreach ((var oldItem, _) in previousTable)
             {
-                bool inItemSet = itemsSet.Remove(oldItem);
-                builder.AddEntriesFromPreviousTable(previousTable, inItemSet ? EntryState.Cached : EntryState.Removed);
+                if (itemsSet.Remove(oldItem))
+                {
+                    // we're iterating the table, so know that it has entries
+                    builder.TryUseCachedEntries();
```

Comment:
Think we should still assert that they are used. 
```suggestion
                    var usedCache = builder.TryUseCachedEntries();
                    Debug.Assert(usedCache);
```